### ISSUE TITLE
Use repo email config for uncommitted changes

### ISSUE THREAD

**Issue by janlarres-msft** [gitkraken/vscode-gitlens#1910]
I have configured Git with my personal email address in the global config, and with my work email address in the local Git config in work repos. This means that committed changes use my work email address and so show correctly in Gitlens. However for uncommitted changes Gitlens seems to look at the global config, and so they show up as being authored using my personal email address, which Gitlens then treats as two different authors.

I know I could probably solve this by adding a `.mailmap` file to the repositories, but I don't think that's really the right solution because it would mean adding my personal email address to work repos just for Gitlens's sake. So would it be possible to make Gitlens take the local repo config into account when showing uncommitted changes?

The same issue probably applies to the configured name, but at least in my case that's identical in both locations.

**Comment by eamodio** [comment]
GitLens *should* be already honoring the "local" config over the global, and also uses any `.mailmap` files to further match. Can you provide more details on what you are seeing? When you say "uncommitted" changes, where are you seeing them? In all the places I can think of GitLens, uncommitted changes are kind of hard-coded to "You".

Also, can you provide the output if you run `git config --get-regex  ^user\.`, since that is what GitLens uses to parse the user and email.

Thanks!

**Comment by janlarres-msft** [comment]
This is what it looks like for uncommitted changes:
![Screen Shot 2022-03-17 at 16 39 26](https://user-images.githubusercontent.com/95390456/158732737-f0b8b7b4-201f-4217-b705-3f49328c1808.png)
The avatar is associated with my personal email address, and that's what Gitlens considers to be "me".

This is what it looks like for committed changes:
![Screen Shot 2022-03-17 at 16 41 39](https://user-images.githubusercontent.com/95390456/158732831-bf1d9847-d90a-457a-ab9c-8e4e6c38d53d.png)
It shows an auto-generated avatar because I don't have my work email address registered with Gravatar. It also doesn't seem to think that this is me and so doesn't say "you".

Running the config command gives me this:
```
$ git config --get-regexp '^user\.'
user.name Jan Larres
user.email <personal email>
user.email <company email>
```

It looks like Git lists global config results first, and I assume that Gitlens always picks the first result.
I noticed that `git config` has an option `--show-scope`, which results in this:
```
$ git config --show-scope --get-regexp '^user\.'
global  user.name Jan Larres
global  user.email <personal email>
local   user.email <company email>
```

This looks like it should be possible to check for a local value first, and if none is found it could fall back to the global value. It would be great if that could be added. Thanks!

**Comment by eamodio** [comment]
I can't seem to reproduce this, and I've dug into this more -- as GitLens picks the last result. It loops through all the values and will overwrite any previous ones found. Which is the behavior I see when I test it out.

Do you have a `.mailmap` on this repo?

After GitLens runs `git config --get-regexp '^user\.'`, it takes the resulting (last one wins) values and creates an "author" string in the form of `{user.name} <{user.email}>` (without the `{}`) and then calls `git check-mailmap <author-string-here>`. So, I wonder if that call is returning your personal e-mail or something.

**Comment by janlarres-msft** [comment]
Okay, I've figured it out. It's actually a bit convoluted.

I'm using a dev container for this project, and I assume that Gitlens gets its information from inside the container if one is used. Now as I mentioned I have configured local settings in my repository, but the settings aren't actually directly in the `.git/config` file but are included with an `[include]` block. And the path in that block points to a file that doesn't get mounted into the container, so when I run the `git config` command it's not able to find the included file. It looks like VS Code does copy the global `~/.config/git/config` file into the container though, and so my personal email address is the only one found when running `git config` inside the container. Previously I had run the Git commands outside the container because there push/pull operations can authenticate easily and so that's my normal way of running Git.

I'm not sure if there's a good solution to this other than setting the email address directly in the local config file or mounting the included directory into the container. I've wanted to avoid the former so the settings are only in place for all repos, and the latter isn't really feasible because the container run configuration is part of the repo setup and so not really user-customizable. In any case this doesn't actually seem to be a Gitlens issue. Thanks for looking into it!
